### ISSUE TITLE
Hub->Run() never quits

### ISSUE THREAD

**Issue by Oipo** [uWebSockets/uWebSockets#572]
## Include the following information (where applicable):

* What operating system are you running?
  * Linux
* Are you running any CPU architechute out of the normal?
  * No, it's 64-bit
* What platform/language are you using?
  * C++
* Are you using SSL, if so, how?
  * SSL is not part of the equation
* What version of uws are you using?
  * master
* What features of µWS are involved in the issue?
  * Server support (WebSocket)
* How does the issue reproduce?
  * All the time
* Can you show me some short code that triggers your issue or any other important information?

In the following program, only "should quit" ever prints, none of the other cout lines are ever reached.

```c++
#include "uWS.h"

#include <iostream>
#include <thread>
#include <chrono>

using namespace std;


int main() {
    try {
        bool quit = false;
        uWS::Hub h;

        auto uws_thread = make_unique<thread>([&h]{
            if(!h.listen(3000)) {
                cout << "h.listen failed" << endl;
                return;
            }
            h.run();
            cout << "Managed to outrun the clutches of uws" << endl;
        });

        this_thread::sleep_for(100ms);

        static_cast<uWS::Group<uWS::SERVER>>(h).terminate();
        cout << "should quit" << endl;
        uws_thread->join();
        cout << "has quit" << endl;
    } catch (std::runtime_error &e) {
        cout << "runtime_error: " << e.what() << endl;
    }

    return 0;
}
```

**Comment by alexhultman** [comment]
You can only terminate from the same thread currently. So you could possibly have some GET path that triggers a terminate or something, or use Async to defer some function to the hub's thread.

**Comment by smasherprog** [comment]
I am also wondering about how to solve this. I tried closing the group, but that only works some of the times. 
Sometimes, it never unblocks, other times it does unblock. It seems to be some sort of race condition.
I am using libuv.
Additionally, I ran this and it never completes https://github.com/uWebSockets/uWebSockets/blob/master/tests/main.cpp#L280

**Comment by alexhultman** [comment]
Well then read what I wrote

**Comment by Oipo** [comment]
This isn't going to be fixed?

**Comment by alexhultman** [comment]
No, it is not an issue. Like I said (and I have said this probably 20 times now, search the issue tracker) you need to use an Async to defer terminate to the thread it belongs to. Any "fix" would really just be syntactic sugar that wraps Async and this is not very highly prioritized. Higher priority is getting better documentation out so that I won't have to repeat myself over and over and over and over again.

**Comment by Oipo** [comment]
#366 

Next time, just link to the issue where you explained it before. Saves both of us time.

Not to mention, it's odd that you close issues that are not actually fixed, even if it's low priority.

**Comment by alexhultman** [comment]
No, it would save YOU time. I don't know the issue number I only know I have answered this before. Instead YOU need to do YOUR OWN research before posting duplicates.

**Comment by socmag** [comment]
Feeling stressed there @alexhultman?   I know the feeling!

But yes @Oipo it isn't a problem, and I agree with Alex, he's only one guy and there is a lot of communication around this project with it being so good and all.  Very hard to keep up although I have to say he is doing pretty damn good.

Maybe prople could fill in the Wiki...
